src/librustc_typeck/mem_categorization.rs

```diff
@@ -679,20 +747,88 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         op(&place_with_id, pat);
 
         match pat.kind {
-            PatKind::TupleStruct(_, ref subpats, _) | PatKind::Tuple(ref subpats, _) => {
-                // S(p1, ..., pN) or (p1, ..., pN)
+            PatKind::Tuple(ref subpats, _) => {
+                // (p1, ..., pN)
                 for subpat in subpats.iter() {
                     let subpat_ty = self.pat_ty_adjusted(&subpat)?;
-                    let sub_place = self.cat_projection(pat, place_with_id.clone(), subpat_ty);
+                    let sub_place = self.cat_projection(
+                        pat,
+                        place_with_id.clone(),
+                        subpat_ty,
+                        ProjectionKind::Other,
+                    );
                     self.cat_pattern_(sub_place, &subpat, op)?;
                 }
             }
 
-            PatKind::Struct(_, field_pats, _) => {
+            PatKind::TupleStruct(ref qpath, ref subpats, dots_pos) => {
+                // S(p1, ..., pN)
+                let variant_index = if let Some(variant_index) =
+                    self.variant_index_for_adt(qpath, pat.hir_id, pat.span)
+                {
+                    variant_index
+                } else {
+                    return Err(());
+                };
+
+                let total_fields = {
+                    let ty = self.tables.node_type(pat.hir_id);
+                    match ty.kind {
+                        ty::Adt(adt_def, _) => adt_def.variants[variant_index].fields.len(),
+                        _ => span_bug!(
```

Comment:
probably should use delayed span here as well, though we will never get here because we would've errored out on L771

Reply:
Maybe extract this as a separate function.